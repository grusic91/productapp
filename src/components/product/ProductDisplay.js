import React from 'react';
import PropTypes from 'prop-types';
import { ProductEditor } from './ProductEditor';
import { ProductTable } from './ProductTable';

export class ProductDisplay extends React.Component {
    constructor(props){
        super(props);
        this.state = {
            showEditor: false,
            selectedProduct: null
        }
    }

    startEditing = product => {
        this.setState({
            showEditor: true,
            selectedProduct: product
        })
    }

    createProduct = () => {
        this.setState({
            showEditor: true,
            selectedProduct: {}
        })
    }

    cancelEditing = () => {
        this.setState({
            showEditor: false,
            selectedProduct: null
        })
    }

    saveProduct = product => {
        this.props.saveCallback(product);
        this.setState({
            showEditor: false,
            selectedProduct: null
        });
    }

    render() {
        const { showEditor, selectedProduct } = this.state;        

        return (
           <div className="m-2">            
            { showEditor ? 
                <div>
                    <ProductEditor
                        key={selectedProduct.id || -1 }
                        product={selectedProduct}
                        saveCallback={this.saveProduct}
                        cancelCallback={this.cancelEditing}
                />
                </div>
                    :
                <React.Fragment>
                    <ProductTable
                        products={this.props.products}
                        editCallback={this.startEditing}
                        deleteCallback={this.props.deleteCallback}
                    />
                    <div className="text-center">
                        <button className="btn btn-primary m-1" onClick={this.createProduct}>
                            Create Product
                        </button>
                    </div>
                </React.Fragment>                 
                }
            </div>      
        )
    }
}

ProductDisplay.propTypes = {
    name: PropTypes.string.isRequired,
    products: PropTypes.array.isRequired,
    saveCallback: PropTypes.func.isRequired,
    deleteCallback: PropTypes.func.isRequired,
}