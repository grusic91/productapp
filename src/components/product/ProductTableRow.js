import React from 'react';

export class ProductTableRow extends React.Component {
    render() {
        let p = this.props.product;
        if(p) {
            return <tr>
            <td>{ p.id }</td>
            <td>{ p.name }</td>
            <td>{ p.category}</td>
            <td className="text-right">${ Number(p.price).toFixed(2) }</td>
            <td>
                <button 
                    className="btn btn-sm btn-warning m-1"
                    onClick={ () => this.props.editCallback(p) }
                > 
                    Edit 
                </button>
                <button
                    className="btn btn-sm btn-danger m-1"
                    onClick={ () => this.props.deleteCallback(p) }
                >
                    Delete
                </button>
            </td>
        </tr>
        }
        
        return <h1>Loading...</h1>
    }
}
