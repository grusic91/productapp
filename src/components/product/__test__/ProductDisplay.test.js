import React from 'react';
import { shallow, mount } from 'enzyme';
import { ProductDisplay } from '../ProductDisplay';
import { ProductTable } from '../ProductTable';
import { ProductEditor } from '../ProductEditor';
import { ProductTableRow } from '../ProductTableRow';


describe('<ProductDisplay />', () => {
    let wrapper;

    beforeAll(() => wrapper = shallow(<ProductDisplay 
        name={""}
        products={[]}
        saveCallback={jest.fn()}
        deleteCallback={jest.fn()}
    />));

    it("should render properly", () => {
        expect(wrapper).toMatchSnapshot();
    });

    it("should render two divs elements", () => {
        expect(wrapper.find('div')).toHaveLength(2);
    });

    it("conditional Rendering ProductTable", () => {
        const productTableProps = wrapper.find(ProductTable).props();
        wrapper.setState({ showEditor: false});
        expect(wrapper.containsMatchingElement(
            <ProductTable {...productTableProps}/>
        )).toEqual(true);
    });
    
    it("conditional Rendering Create Product <button> element", () => {
        wrapper.setState({ showEditor: false});
        expect(wrapper.find('button').text()).toEqual("Create Product");
    });

    it("conditional Rendering ProductEditor", () => {
        wrapper.setState({ 
            showEditor: true, 
            selectedProduct: {
                id: "",
                name: "",
                category: "",
                price: ""
            }
        });
        expect(wrapper.containsMatchingElement(<ProductEditor />)).toEqual(true);
    });
});

describe('<ProductDisplay /> functionality', () => {
    let wrapper;

    const saveData = (collection, item) => {
        if (item.id === "") {
            item.id = this.idCounter++;
            this.setState(state => state[collection] = state[collection].concat(item))
        } else {
            this.setState(state => state[collection] = state[collection].map(
                stored => stored.id === item.id ? item: stored
            ))
        }
    }

    const deleteData = (collection, item) => {
        this.setState(state => state[collection] = state[collection].filter(
            stored => stored.id !== item.id
        ));
    }

    beforeAll(() => wrapper = mount(<ProductDisplay 
        name={"Products"}
        products={[
            {
                id: 1,
                name: "Kayak1",
                category: "Watersports",
                price: "275"
            },
            {
                id: 2,
                name: "Lifejacket",
                category: "Watersports",
                price: "48.95"
            }
        ]}
        saveCallback={saveData}
        deleteCallback={deleteData}
    />));
    
    it("click 'Edit' button in product table, ProductEditor component should be rendered", () => {
        const productTableComponent = wrapper.find(ProductTable);
        const editButton = productTableComponent.find(ProductTableRow).at(0).find('button').at(0);
        editButton.simulate('click');

        expect(wrapper.containsMatchingElement(ProductEditor)).toEqual(true);
    });
});

