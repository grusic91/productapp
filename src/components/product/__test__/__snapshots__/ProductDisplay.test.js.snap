// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductDisplay /> should render properly 1`] = `ShallowWrapper {}`;
